// @stacker/alfred-utils/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.input = exports.WorkflowUpdater = exports.Workflow = exports.dateUtils = exports.utils = exports.http = void 0;
var http_1 = require("./http");
Object.defineProperty(exports, "http", { enumerable: true, get: function () { return __importDefault(http_1).default; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "utils", { enumerable: true, get: function () { return __importDefault(utils_1).default; } });
var date_utils_1 = require("./date-utils");
Object.defineProperty(exports, "dateUtils", { enumerable: true, get: function () { return __importDefault(date_utils_1).default; } });
var workflow_1 = require("./workflow");
Object.defineProperty(exports, "Workflow", { enumerable: true, get: function () { return workflow_1.Workflow; } });
var workflow_updater_1 = require("./workflow-updater");
Object.defineProperty(exports, "WorkflowUpdater", { enumerable: true, get: function () { return workflow_updater_1.WorkflowUpdater; } });
/**
 * @description 读取命令参数
 */
exports.input = process.argv[2];
